#### Bernoulli Naive Bayes

It is a kind of naive bayes algorithm that can work with the binary values (0 and 1).

It is mainly used for binary classification.

#### Step 1:    Import  all the necessary libraries

In [1507]:
import  numpy               as  np
import  pandas              as  pd
import  matplotlib.pyplot   as  plt
import  seaborn             as  sns

from    sklearn.model_selection          import  train_test_split, GridSearchCV, StratifiedKFold
from    sklearn.preprocessing            import  StandardScaler
from    sklearn.feature_extraction.text  import  CountVectorizer
from    sklearn.naive_bayes              import  BernoulliNB
from    sklearn.metrics                  import  accuracy_score, confusion_matrix, classification_report

#### Step 2: Load the dataset

In [1508]:
df = pd.read_csv("spam.csv", encoding = 'latin-1')

df.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


#### OBSERVATIONS:

1.  The above dataset contains the email text and it's type.

2.  The independent feature is the email text.

3.  The dependent feature is the type of the text.

#### Step 3:    Perform the Data Exploration

In [1509]:
#### get the top five rows of the dataset

df.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


In [1510]:
#### get the bottom five rows of the dataset

df.tail()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
5567,spam,This is the 2nd time we have tried 2 contact u...,NaN,NaN,NaN
5568,ham,Will Ì_ b going to esplanade fr home?,NaN,NaN,NaN
5569,ham,"Pity, * was in mood for that. So...any other s...",NaN,NaN,NaN
5570,ham,The guy did some bitching but I acted like i'd...,NaN,NaN,NaN
5571,ham,Rofl. Its true to its name,NaN,NaN,NaN


In [1511]:
#### get the length of the dataset

print("Length of the dataset is:", len(df))

Length of the dataset is: 5572


In [1512]:
#### get the shape of the dataset

print("Shape of the dataset is:", df.shape)

Shape of the dataset is: (5572, 5)


In [1513]:
#### get the columns used in the dataset

df.columns

Index(['v1', 'v2', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'], dtype='object')

In [1514]:
#### get the information about the columns used in the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


In [1515]:
#### get the descriptive summary statistics about the columns used in the dataset

df.describe()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
count,5572,5572,50,12,6
unique,2,5169,43,10,5
top,ham,"Sorry, I'll call later","bt not his girlfrnd... G o o d n i g h t . . .@""","MK17 92H. 450Ppw 16""","GNT:-)"""
freq,4825,30,3,2,2


#### Step 4: Perform Feature Engineering and Data Cleaning

In [1516]:
#### Remove all the unnecessary columns from the dataset

df.drop(columns={"Unnamed: 2","Unnamed: 3","Unnamed: 4"},inplace=True)

In [1517]:
df.head()

,v1,v2
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [1518]:
#### Reset the columns to their proper position

df = df[['v2','v1']]

In [1519]:
df.head()

,v2,v1
0,"Go until jurong point, crazy.. Available only ...",ham
1,Ok lar... Joking wif u oni...,ham
2,Free entry in 2 a wkly comp to win FA Cup fina...,spam
3,U dun say so early hor... U c already then say...,ham
4,"Nah I don't think he goes to usf, he lives aro...",ham


In [1520]:
#### Rename the columns to the proper names

df.rename(columns={"v2":"text", "v1":"label"},inplace=True)

In [1521]:
df.head()

,text,label
0,"Go until jurong point, crazy.. Available only ...",ham
1,Ok lar... Joking wif u oni...,ham
2,Free entry in 2 a wkly comp to win FA Cup fina...,spam
3,U dun say so early hor... U c already then say...,ham
4,"Nah I don't think he goes to usf, he lives aro...",ham


In [1522]:
#### Map the "ham" and "spam" to their proper numbers

from  sklearn.preprocessing  import  LabelEncoder

#### create an object for label encoder

label = LabelEncoder()

df['label'] = label.fit_transform(df['label'])

#df['label'] = df['label'].map({"spam":1,"ham":0})

In [1523]:
df.head()

,text,label
0,"Go until jurong point, crazy.. Available only ...",0
1,Ok lar... Joking wif u oni...,0
2,Free entry in 2 a wkly comp to win FA Cup fina...,1
3,U dun say so early hor... U c already then say...,0
4,"Nah I don't think he goes to usf, he lives aro...",0


#### OBSERVATIONS:

1.  The "ham" and "spam" label has been converted to their proper numbers 0 and 1 respectively.

In [1524]:
#### To check for any NULL record in the dataset

df.isnull().sum()

text     0
label    0
dtype: int64

#### OBSERVATIONS:

1.  There are no NULL record in the dataset.

In [1525]:
#### To check for the duplicate record in the dataset

df[df.duplicated()]

,text,label
102,As per your request 'Melle Melle (Oru Minnamin...,0
153,As per your request 'Melle Melle (Oru Minnamin...,0
206,"As I entered my cabin my PA said, '' Happy B'd...",0
222,"Sorry, I'll call later",0
325,No calls..messages..missed calls,0
...,...,...
5524,You are awarded a SiPix Digital Camera! call 0...,1
5535,"I know you are thinkin malaria. But relax, chi...",0
5539,Just sleeping..and surfing,0
5553,Hahaha..use your brain dear,0


#### OBSERVATIONS:

1.  There are 403 duplicate records in the dataset. So remove it from the dataset.

In [1526]:
#### Remove the duplicates from the datasey

df = df.drop_duplicates()

In [1527]:
#### To again check for the duplicate record in the dataset

df[df.duplicated()]

,text,label


#### OBSERVATIONS:

1.  There are no duplicate records in the dataset.

In [1528]:
#### Again check for the length of the dataset

print("Length of the revised dataset is:", len(df))

Length of the revised dataset is: 5169


In [1529]:
#### Again check for the shape of the new dataset

print("Shape of the revised dataset is:", df.shape)

Shape of the revised dataset is: (5169, 2)


In [1530]:
#### Check the categories in the output label

df['label'].value_counts()

label
0    4516
1     653
Name: count, dtype: int64

#### Step 5: Divide the dataset into independent and dependent variables

In [1531]:
#### independent variables

X = df['text']

print(X)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                Will Ì_ b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: text, Length: 5169, dtype: object


In [1532]:
#### dependent variables

Y = df['label']

In [1533]:
print(Y)

0       0
1       0
2       1
3       0
4       0
       ..
5567    1
5568    0
5569    0
5570    0
5571    0
Name: label, Length: 5169, dtype: int64


#### Step 6: Divide the independent and dependent variables into training and testing data

In [1534]:
from  sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,random_state=42,stratify=Y)

In [1535]:
X_train

336     Cool. So how come you havent been wined and di...
390                             Huh so late... Fr dinner?
583     my ex-wife was not able to have kids. Do you w...
1387    Hope ur head doesn't hurt 2 much ! Am ploughin...
1095                              Ryder unsold.now gibbs.
                              ...                        
606     XCLUSIVE@CLUBSAISAI 2MOROW 28/5 SOIREE SPECIAL...
105     Thanks a lot for your wishes on my birthday. T...
3789    I love you !!! You know? Can you feel it? Does...
1911    Lol grr my mom is taking forever with my presc...
3742    Nah, Wednesday. When should I bring the mini c...
Name: text, Length: 4135, dtype: object

In [1536]:
X_test

5318    Good morning, my Love ... I go to sleep now an...
5281               And how you will do that, princess? :)
2459       Nope. Meanwhile she talk say make i greet you.
1305    Designation is software developer and may be s...
5083    Aiya we discuss later lar... Pick Ì_ up at 4 i...
                              ...                        
1119    Bored of speed dating? Try SPEEDCHAT, txt SPEE...
3068                   Ay wana meet on sat?Ì_ wkg on sat?
812     Congratulations ur awarded either å£500 of CD ...
1560    Just got some gas money, any chance you and th...
1238    Dear relieved of westonzoyland, all going to p...
Name: text, Length: 1034, dtype: object

In [1537]:
print("Shape of the input training data is:", X_train.shape)
print("Shape of the input testing  data is:", X_test.shape)

Shape of the input training data is: (4135,)
Shape of the input testing  data is: (1034,)


In [1538]:
Y_train

336     0
390     0
583     0
1387    0
1095    0
       ..
606     1
105     0
3789    0
1911    0
3742    0
Name: label, Length: 4135, dtype: int64

In [1539]:
Y_test

5318    0
5281    0
2459    0
1305    0
5083    0
       ..
1119    1
3068    0
812     1
1560    0
1238    0
Name: label, Length: 1034, dtype: int64

In [1540]:
print("Shape of the output training data is:", Y_train.shape)
print("Shape of the output testing  data is:", Y_test.shape)

Shape of the output training data is: (4135,)
Shape of the output testing  data is: (1034,)


#### Step 7: Perform Count Vectorization on the inputs

In [1541]:
from  sklearn.feature_extraction.text  import  CountVectorizer

###3 create an object for CountVectorizer

count = CountVectorizer(stop_words='english')

#### using the object of count vectorizer, transform the inputs

X_train_vectorized = count.fit_transform(X_train)

print(X_train_vectorized)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 31574 stored elements and shape (4135, 7329)>
  Coords	Values
  (0, 1928)	1
  (0, 1831)	1
  (0, 3212)	1
  (0, 7113)	1
  (0, 2239)	1
  (1, 3391)	1
  (1, 3850)	1
  (1, 2840)	1
  (1, 2243)	1
  (2, 2571)	1
  (2, 7096)	1
  (2, 749)	1
  (2, 3760)	2
  (2, 6984)	1
  (2, 2097)	1
  (3, 1831)	1
  (3, 3342)	1
  (3, 6821)	1
  (3, 3221)	1
  (3, 2297)	1
  (3, 3407)	1
  (3, 4990)	1
  (3, 7016)	1
  (3, 4952)	1
  (3, 3575)	1
  :	:
  (4132, 2295)	2
  (4132, 5755)	1
  (4132, 3776)	1
  (4132, 812)	1
  (4132, 6991)	1
  (4132, 2352)	1
  (4132, 1253)	1
  (4132, 856)	1
  (4133, 3932)	1
  (4133, 4005)	1
  (4133, 6350)	1
  (4133, 2813)	1
  (4133, 1131)	1
  (4133, 6744)	1
  (4133, 4307)	1
  (4133, 4367)	1
  (4133, 3120)	1
  (4133, 5113)	1
  (4133, 4917)	1
  (4134, 1444)	1
  (4134, 7037)	1
  (4134, 1162)	1
  (4134, 4470)	1
  (4134, 4296)	1
  (4134, 1689)	1


In [1542]:
X_test_vectorized = count.transform(X_test)

print(X_test_vectorized)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 6868 stored elements and shape (1034, 7329)>
  Coords	Values
  (0, 1149)	1
  (0, 1271)	1
  (0, 2097)	1
  (0, 2688)	1
  (0, 3048)	1
  (0, 3099)	1
  (0, 3776)	1
  (0, 4046)	2
  (0, 4389)	1
  (0, 4723)	1
  (0, 5899)	1
  (0, 6506)	1
  (0, 7125)	1
  (1, 5135)	1
  (2, 3104)	1
  (2, 4137)	1
  (2, 4591)	1
  (2, 5628)	1
  (2, 6353)	1
  (3, 1691)	1
  (3, 5968)	1
  (4, 867)	1
  (4, 2265)	1
  (4, 3843)	1
  (4, 3853)	1
  :	:
  (1031, 2348)	1
  (1031, 2508)	1
  (1031, 2853)	1
  (1031, 2996)	1
  (1031, 3870)	1
  (1031, 4451)	1
  (1031, 6570)	1
  (1031, 6722)	1
  (1031, 6821)	1
  (1031, 6934)	1
  (1031, 7044)	1
  (1031, 7217)	1
  (1032, 1647)	1
  (1032, 2938)	1
  (1032, 2946)	1
  (1032, 3064)	1
  (1032, 3085)	1
  (1032, 3707)	1
  (1032, 4374)	1
  (1032, 4495)	1
  (1032, 6984)	1
  (1033, 2107)	1
  (1033, 2479)	1
  (1033, 3035)	1
  (1033, 4969)	1


#### OBSERVATIONS:

1.  The input training and testing data which was in text form has been converted into the matrix of word count.

#### Step 8: Train the Bernoulli Naive Bayes Model

In [1543]:
from  sklearn.naive_bayes  import  BernoulliNB

#### create an object for Bernoulli Naive Bayes

bnb = BernoulliNB()

#### using the object of Bernoulli NB, train the model

bnb.fit(X_train_vectorized, Y_train)

BernoulliNB()

#### OBSERVATIONS:

1.  The Bernoulli Naive Bayes Model has been trained with the training data.

#### Step 9: Predict the Model

In [1544]:
Y_pred = bnb.predict(X_test_vectorized)

In [1545]:
Y_pred

array([0, 0, 0, ..., 1, 0, 0])

#### Step 10: Perform the Classification Evaluation Metrics

In [1546]:
from        sklearn.metrics             import  accuracy_score, confusion_matrix, classification_report

print("===============================Accuracy score=======================================================")

ac = accuracy_score(Y_test, Y_pred)*100.0

print("Accuracy score of the model is:", ac)



print("===============================Confusion matrix=======================================================")

cm = confusion_matrix(Y_test, Y_pred)

print("Confusion matrix of the model is:", cm)



print("===============================Classification report=======================================================")

cr = classification_report(Y_test, Y_pred)

print("Classification report of the model is:", cr)

===============================Accuracy score=======================================================
Accuracy score of the model is: 96.32495164410058
===============================Confusion matrix=======================================================
Confusion matrix of the model is: [[903   0]
 [ 38  93]]
===============================Classification report=======================================================
Classification report of the model is:               precision    recall  f1-score   support

           0       0.96      1.00      0.98       903
           1       1.00      0.71      0.83       131

    accuracy                           0.96      1034
   macro avg       0.98      0.85      0.90      1034
weighted avg       0.96      0.96      0.96      1034



#### OBSERVATIONS:

1.  The accuracy of the Bernoulli Naive Bayes model using spam classification dataset is 96.32

2.  The accuracy of the model is very high. So the model is working fine.

#### Step 11: Predict whether the email is spam or ham based on the sample data

In [1547]:
#### define the sample messages

sample_messages = [
    "Congratulations! You've won a free iPhone. Click here to claim.",
    "Hey, are we still meeting for lunch today?"
]

print(sample_messages)


#### perform vectorization on the inputs

vectorized_messages = count.transform(sample_messages)

print(vectorized_messages)


#### perform the prediction on the inputs

prediction = bnb.predict(vectorized_messages)

print(prediction)



print("==========================First Sentence===============================================")

if(prediction[0] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")


print("==========================Second Sentence===============================================")

if(prediction[1] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")

["Congratulations! You've won a free iPhone. Click here to claim.", 'Hey, are we still meeting for lunch today?']
<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 11 stored elements and shape (2, 7329)>
  Coords	Values
  (0, 1748)	1
  (0, 1766)	1
  (0, 1888)	1
  (0, 2853)	1
  (0, 3569)	1
  (0, 6879)	1
  (0, 7157)	1
  (1, 3273)	1
  (1, 4075)	1
  (1, 4236)	1
  (1, 6576)	1
[0 0]
==========================First Sentence===============================================
The email message is Ham
==========================Second Sentence===============================================
The email message is Ham


#### Step 12: Perform Hyper Parameter Tuning using GridSearchCV

In [1548]:
#### define the pipeline

from  sklearn.pipeline  import  Pipeline

#### create an object for pipeline

pipeline = Pipeline([
    ("count"    ,   CountVectorizer())      ,
    ("bnb"      ,   BernoulliNB())
])

In [1549]:
#### create the param grid

param_grid = {
     "count__ngram_range" : [(1,1), (1,2)]              ,
     "count__max_df"      : [0.75, 1.0]                 ,
     "count__min_df"      : [1, 2]                      ,
     "count__stop_words"  : [None, 'english']           ,
     "count__binary"      : [True]                      ,
     "bnb__alpha"         : [0.01, 0.1, 1.0]            ,
     "bnb__fit_prior"      : [True, False]
}

In [1550]:
#### perform the Cross Validation 

from  sklearn.model_selection  import  StratifiedKFold

cv = StratifiedKFold(n_splits = 5, shuffle = True, random_state=42)

In [1551]:
#### create the object for grid

grid = GridSearchCV(
    pipeline                                            ,
    param_grid       =       param_grid                 ,
    cv               =           cv                     ,
    scoring          =       'accuracy'                 ,
    n_jobs           =         -1
)

In [1552]:
#### Train the model

grid.fit(X_train, Y_train)

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=42, shuffle=True),
             estimator=Pipeline(steps=[('count', CountVectorizer()),
                                       ('bnb', BernoulliNB())]),
             n_jobs=-1,
             param_grid={'bnb__alpha': [0.01, 0.1, 1.0],
                         'bnb__fit_prior': [True, False],
                         'count__binary': [True], 'count__max_df': [0.75, 1.0],
                         'count__min_df': [1, 2],
                         'count__ngram_range': [(1, 1), (1, 2)],
                         'count__stop_words': [None, 'english']},
             scoring='accuracy')

In [1553]:
#### Obtain the best parameters and the estimator for the model

best_parameter = grid.best_params_

print("Best Parameters for the model is:", best_parameter)


best_estimator = grid.best_estimator_

print("Best Estimator for the model is:", best_estimator)

Best Parameters for the model is: {'bnb__alpha': 0.1, 'bnb__fit_prior': False, 'count__binary': True, 'count__max_df': 0.75, 'count__min_df': 1, 'count__ngram_range': (1, 1), 'count__stop_words': None}
Best Estimator for the model is: Pipeline(steps=[('count', CountVectorizer(binary=True, max_df=0.75)),
                ('bnb', BernoulliNB(alpha=0.1, fit_prior=False))])


In [1554]:
#### predict the model

Y_pred_grid = best_estimator.predict(X_test)

In [1555]:
Y_pred_grid

array([0, 0, 0, ..., 1, 0, 0])

In [1556]:
#### Evaluate the Classification metrics

from        sklearn.metrics             import  accuracy_score, confusion_matrix, classification_report

print("===============================Accuracy score=======================================================")

acgrid = accuracy_score(Y_test, Y_pred_grid)*100.0

print("Accuracy score of the model is:", acgrid)



print("===============================Confusion matrix=======================================================")

cmgrid = confusion_matrix(Y_test, Y_pred_grid)

print("Confusion matrix of the model is:", cmgrid)



print("===============================Classification report=======================================================")

crgrid = classification_report(Y_test, Y_pred_grid)

print("Classification report of the model is:", crgrid)

===============================Accuracy score=======================================================
Accuracy score of the model is: 98.8394584139265
===============================Confusion matrix=======================================================
Confusion matrix of the model is: [[901   2]
 [ 10 121]]
===============================Classification report=======================================================
Classification report of the model is:               precision    recall  f1-score   support

           0       0.99      1.00      0.99       903
           1       0.98      0.92      0.95       131

    accuracy                           0.99      1034
   macro avg       0.99      0.96      0.97      1034
weighted avg       0.99      0.99      0.99      1034



#### OBSERVATIONS:

1.  The accuracy of the Berboulli Naive Bayes model for email spam classification has improved from 6.32 till 98.83.

2.  The accuracy of the model has increased. So it is working fine.

#### Step 13: Predict whether the text is "ham" or "spam" based on the new test data

In [1557]:
#### define the sample messages

sample_messages = [
    "Congratulations! You've won a free iPhone. Click here to claim.",
    "Hey, are we still meeting for lunch today?"
]

print(sample_messages)


#### perform the prediction on the vectorized data

prediction = best_estimator.predict(sample_messages)

print(prediction)


print("==========================First Sentence===============================================")

if(prediction[0] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")


print("==========================Second Sentence===============================================")

if(prediction[1] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")

["Congratulations! You've won a free iPhone. Click here to claim.", 'Hey, are we still meeting for lunch today?']
[1 0]
==========================First Sentence===============================================
The email message is Spam
==========================Second Sentence===============================================
The email message is Ham
